### DISCUSSION TITLE
How to set path parameter

### DISCUSSION THREAD

**Discussion by mirasuba** [mockoon/mockoon#487] [Help] [answered]
Hi, 

pls, i need help with mock one of my api.

I will try to explaine what i need. :) 

I have existing api but without function swagger so i need to create mock in mockoon without definition.

The target request looks like this: http://localhost:3013/test/Processes(ExternalId='{1002000100}') .

When i have it like this in mockoon:
![image](https://user-images.githubusercontent.com/84851842/119668410-74fabc00-be37-11eb-858a-0a5792520916.png)

I get "Cannot GET /test/Processes(ExternalId=&#39;%7B1002000100D&#39;)"

Do you know someone how to set it up? I tried to sets ExternalId as a query or route param but without succes. ExternalID is in api definition is like path parameter.

Thank you,
m.

**Comment by 255kb** [comment, marked as answer]
Hi, you cannot declare paths with characters like `()` as they are interpreted similarly to their regex counterpart. See [ExpressJS routing documentation](https://expressjs.com/en/guide/routing.html) for more info on this (ExpressJS is the library used by Mockoon to create the mocks).

If you want to pass parameters to Mockoon, you can either use the path parameter syntax:
`path/:id`
You can then access/use the `id` in the body with `{{urlParam 'id'}}` or in the rules by selecting "Route params".

Or use the query parameters, by declaring a route without them like `path` and calling the mock endpoint with the query `path?id=123`. 
You can then access/use the `id` in the body with `{{queryParam 'id'}}` or in the rules by selecting "Query string".

**Reply by JaNeProgrammist** [reply]
Hi @255kb , thank you for your answer, I would like to add some more details to this issue. 
I have a problem with the PATCH method. I have a few parameters, each of which includes brackets. Lets say I have 3 such methods with different parameters each:
![image](https://user-images.githubusercontent.com/60709524/119964341-66ccad00-bfa9-11eb-9037-d5a53cab49ee.png)
I put my parameters into the "rules" section like this, since I cant put them right into the path:
![image](https://user-images.githubusercontent.com/60709524/119964439-819f2180-bfa9-11eb-962a-71d71495ead9.png)
Obviously, I do this for all 3 calls same way. All of them have different responses which look like this: 
![image](https://user-images.githubusercontent.com/60709524/119964759-d3e04280-bfa9-11eb-8731-6a2040ca9bed.png)

But the problem is, no matter which parameter I use to call, I will always get the same response.
For example, I call PATCH 'localhost:3003/bla/Number('{id}')' but I get the response 
{
      "param1": "This is the response for Partners('{id}')"
}
which is, obviously, not the right response. 
![image](https://user-images.githubusercontent.com/60709524/119965834-fde63480-bfaa-11eb-9442-5965821d7fae.png)
![image](https://user-images.githubusercontent.com/60709524/119965882-09396000-bfab-11eb-9e78-1aa15199ab31.png)

The same will happen with the PATCH 'localhost:3003/bla/Address('{id}')' - i will reseve the response for Partners('{id}'). 

Absolutely the same problem occurs if I put all the requests together into one method like this: 
![image](https://user-images.githubusercontent.com/60709524/119965270-5d901000-bfaa-11eb-89fe-7c61b67a789e.png)
Would you, please, help me solve this problem and maybe even offer a solution? 

Thank you
Rami

**Reply by 255kb** [reply]
Hi, 
Concerning the first problem, all your routes are the same. Sure the parameter name will be different behind the scene, but the first route (order is important!) will take precedence and act like a catch all. You can even read your first route `/:param1` like this `/*`. 
So, what you are doing with the fourth route `/hello/:param` is the correct approach. However, you need to serve the different responses based on something. With only one variable parameter this is quite hard, it could be base on some sort of rules with a regex `^Partners`. 
What I would rather do is to create three routes `/partners/:id`, `/addresses/:id`, `/numbers/:id`. Or one with two parameters: `/:entity/:id`.

**Reply by JaNeProgrammist** [reply]
Hi,
Thank you for your quick answer! I really appreciate your work!

According to your answer I created three routes and put the ('{id}')as a parameter. It seems to work fine, so I think I may leave it like this.
![image](https://user-images.githubusercontent.com/60709524/120151759-f0b88800-c1ec-11eb-99ed-94d153714c94.png)

Unfortunately, such an easy way does not work for the problem when I have the very strict route. 
For example, here I have two get methods, they both have the same route:
/Check(Supply='{8591824006052}'),
/Check(Supply='{27ZG500Z028}').
The only difference in these data is inside the curly brackets.

![image](https://user-images.githubusercontent.com/60709524/120154404-21e68780-c1f0-11eb-98de-26cfe2ce1c0e.png)

I tried to apply the same same trick here as I used above, but, unfortunately, I face the same problem as usual: I get the same response (response for the first request)

What can I do to make it work and show Mockoon the difference between these two routes? Is it possible at all?
As was previously said, I cannot add, delete or change anything in these routes since that should be the only url used in my task.

Thank you again for your help

Rami

**Reply by JaNeProgrammist** [reply]
Okay, @255kb just resolved it myself by accident :D If i chose "Route params" instead of the "Query string" it works just fine :D

**Reply by 255kb** [reply]
Glad you solved it! Indeed, you are using route parameters here `:param` and not query parameters that are separated by a `?`, like in `/path?queryparam1=value1`
